### ISSUE TITLE
For the LUKS2 encrypted home directory use case, allow users to specify `--fs-type=custom`  and supply user-provided mount and umount functions

### ISSUE THREAD

**Issue by jkool702** [systemd/systemd#24371]
### Component

homectl, systemd-homed, userdbctl

### Is your feature request related to a problem? Please describe

The "LUKS2 encryption" use case for systemd-homed looks to be a fantastic feature...The "automatic decryption when a user logs on" and "automatic changing the decryption password when the user password changes" aspects really make it easier to use an encrypted home directory, and the "automatic locking on signout" is a great security feature.

One aspect it is lacking, however, is broad filesystem support - you can only use it with btrfs, ext4 or xfs. This prevents people like me who are thoroughly invested in another filesystem (zfs in my case) from using this fantastic feature.

I understand that there are some good reasons for limited filesystem support (limited dev resources, license issues, etc.). I would be more than willing to self-support a not-officially-supported filesystem, but AFAIK there is currently no way to do this. As such, I propose that you implement a framework to let users who wish to use other filesystems do so in a self-supported way by providing mount/umount commands for systemd-homed to run. This opens up the door for these users to utilize systemd-homed without systemd supporting these filesystems explicitly (and without the time and license issues that would come with it).

### Describe the solution you'd like

The solution I propose involves adding a new `custom` option to the `--fs-type=` parameter.  If a user gives `--fs-type=custom`, systemd-homed would search a pre-defined file/directory list (say, `/etc/systemd/homed/custom-fs-type/control.sh`) and source it. This file would need to contain the following:

```
#!/bin/sh

# the name of the directory under /home that will be mounted/umounted is passed to these functions

mount.custom() {
    # provide custom mount command(s) here

    # example: zfs, where the pool name is the same as the directory name under /home and the dataset is the top level dataset
    zpool import -N "${1}"
    zfs mount -o mountpoint="/home/${1}" "${1}"
}

umount.custom() {    
    # provide custom umount command(s) here

    # example: zfs, where the pool name is the same as the directory name under /home and the dataset is the top level dataset
    zfs umount "${1}"
    zpool export "${1}"
}

```

systemd-homed would then proceed with everything (stuff related to LUKS encryption, managing user credentials, etc.) as it normally would, with 1 small change -  when mounting the home dir (after LUKS decryption or umounting the home dir (before LKS locking), rather than calling the built-in mount/umount commands it would call 

    mount.custom <homedir>

or

    umount.custom <homedir>

Id imagine this would require minimal modification to the code, since almost all of the systemd-homed code is still being used. It would be up to the user to provide and troubleshoot these `mount.custom` and `umount.custom` functions.

### Describe alternatives you've considered

There is currently really no other good way to automatically decrypt (on successful login) and lock (on logout) a home directory on Linux. Some basic functionality (that is buggy and can easily lead to you being locked out) can be achieved using PAM, but nothing else even comes close to systemd-homed.

### The systemd version you checked that didn't have the feature you are asking for

Current version used on Fedora 36

**Comment by poettering** [comment]
I am really not convinced about this. We need a fairly modern fs and knowledge about its feature set, so that we can issue resizes, discards, userns support and so on. Plugging in arbitrary stuff without homed grokking it cannot work.

And I see no avenue for zfs here, sorry. It's not in the kernel and has no perspective to be merged due to the licensing mess, hence this simply won't work out I fear.

Sorry if that's disappointing, but I am pretty sure we don't want this in homed.

We can revisit this if zfs manages to get merged, and has proper ioctls for everything we need and idmapping. But that's a steep request so don't hold your breath.

Sorry.

**Comment by jkool702** [comment]
I appreciate the response. Those are fair reasons.

 In the month since I made this request Ive been working on a way to implement ZFS support with a somewhat different approach. Its mostly working (just polishing off a few aspects), though If you wouldn't mind I had a couple questions for you.

The "short version" of what I am doing is letting homed manage a LUKS2/btrfs home directory that uses a ZFS volume (i.e. ZFS block device) as the location for the LUKS image (so it isnt on a file on disk). Catch is this home directory is almost empty - it only has the `.identity` file and a keyfile. Once homed activated the home area, a script gets triggered (currently using `dbus-monitor` run in a systemd service that looks for the "ActivateHome" method call, though im looking into perhaps using systemd via udev instead). This script loads the keyfile into an encrypted ZFS dataset, unmounts the homed-mounted home directory, and mounts the(now decrypted) ZFS home directory. It also turns out that since homed (seemingly) uses system `umount` commands it will successfully umount the zfs home directory, so long as it is the only thing mounted at/under /home/$USER

My questions are: 

1. How does homed interact with the .identity file? In particular, will it be a problem if its not being present on the filesystem at `/home/$USER/.identity (though still available in the BTRFS filesystem in the `/dev/mapper/home-$USER block device). I tried bind-mounting it on top of the zfs home directory, and homed really didnt like it, though it seems to be fairly happy with it not mounted or present on the filesystem at all.

2.  I understand not wanting to add custom mounting commands, since homed needs particular filesystems to work, but what about adding a sort of `rc.local` file? (think of it like a generalized `.bashrc` that is shell agnostic, lives in the .identity file, and will get run by whatever shell you tell it to use). This would make triggering my remounting script at the right time a good deal easier, but it also seems like it would be useful for general users of homed and doesnt require homed interfacing with any unsupported filesystems.

I *really* like the idea of homed in its "luks" mode, Im just too invested in zfs to not use it on my home dir (e.g., automatic system backups by sending snapsots to a local raidz2 pool with 80 TB raw disk capacity)
